src/validation.cpp

```diff
@@ -4629,7 +4647,14 @@ void Chainstate::LoadExternalBlockFile(
                 // Activate the genesis block so normal node progress can continue
                 if (hash == params.GetConsensus().hashGenesisBlock) {
                     BlockValidationState state;
-                    if (!ActivateBestChain(state, nullptr)) {
+                    bool genesis_activation_failure = false;
+                    for (auto c : GetAll()) {
```

Comment:
534b6f99a3779465678f39ed2704da6049c6469d Previously `LoadExternalBlockFile` was only called on the `ActiveChainstate()`, now you're looping over both. This is ok? If possible, it would be good to move the change to `LoadExternalBlockFile` into a separate commit.

Reply:
If you're asking if it's ok to call `ActivateBestChain()` on both chainstates, the answer to that is essentially always "yes".  ABC() doesn't do anything if no new candidates for most-work chain have been added to our block index; but if any block that has more work than our tip is now available we want to try to connect it.

When calling `LoadExternalBlockFile()`, we're adding blocks that theoretically could be of interest to either chainstate; so I think it makes sense to separate the block storage and import from any particular chainstate.

I'm not sure how easy this would be to separate out into its own commit, because this invokes `AcceptBlock()` as well...  Hopefully the commit is more manageable now that I've split a bunch of other stuff out?  Let me know if you see a good way to shrink this commit more and I can take a stab at it.

Reply:
In commit "Move block-storage-related logic to ChainstateManager" (034f920d1755cff752d46b089ad37bdd703bf896) in `LoadExternalBlockFile`:

It seems inconsistent to call `ActivateBestChain` on _all_ chainstates here for the genesis block, but only call `ActivateBestChain` on the _active_ chainstate a few lines below (line 4631).

I think I can understand the reasons from the discussion above ("we're adding blocks that theoretically could be of interest to either chainstate; so I think it makes sense to separate the block storage and import from any particular chainstate") and from pruning context below. But in both cases I think the code would be clearer if there were explicit comments saying why `ActiveChainstate()` is used and why `GetAll()` is used in each place.